### PR TITLE
Add `ServeFile` and `ServeDir`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Service that serves a file and guesses the mime type with https://crates.io/crates/mime_guess.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
@tesaguri Is this something you would be up for reviewing?

File: tower-http/src/services/fs/serve_dir.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                if append_index_html_on_directories {
+                    let is_dir = tokio::fs::metadata(full_path.clone())
+                        .await
+                        .map(|m| m.is_dir())
+                        .unwrap_or(false);
+
+                    if is_dir {
+                        full_path.push("index.html");
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply tower-http/src/services/fs/serve_dir.rs]: Uh thats a good point. Didn't consider that. Fixed in 467dd06dbaa99

**Review by LucioFranco — APPROVED** [review]
LGTM just one question

File: tower-http/src/services/fs/mod.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+impl<T> AsyncReadBody<T> {
+    /// Create a new [`AsyncReadBody`] wrapping the given reader.
+    fn new(read: T) -> Self {
+        Self { inner: read }
+    }
+}
+
+impl<T> Body for AsyncReadBody<T>
+where
+    T: AsyncRead + Unpin,
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply tower-http/src/services/fs/mod.rs]: Was necessary to do `Pin::new(&mut self.inner)` but adding `pin_project` is easy and gets rid of the bound. I've done that in https://github.com/tower-rs/tower-http/pull/45/commits/6e08550a884ddb8810ea17cc111f3c56674accee.

**Pull request merged**